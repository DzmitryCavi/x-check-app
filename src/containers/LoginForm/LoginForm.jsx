import React from 'react'
import PropTypes from 'prop-types'
import { Redirect } from 'react-router-dom'
import { connect } from 'react-redux'

import { Card, Button, Select, Alert } from 'antd'
import { GithubOutlined } from '@ant-design/icons'

import { login } from '../../store/actions'

import './LoginForm.scss'

class LoginForm extends React.Component {
  constructor(props) {
    super(props)

    this.roles = [
      { slug: 'author', label: 'Author' },
      { slug: 'student', label: 'Student' },
      { slug: 'supervisor', label: 'Supervisor' },
      { slug: 'course_manager', label: 'Course Manager' },
      { slug: 'superadmin', label: 'Super Admin' },
    ]

    this.state = {
      role: localStorage.getItem('role') || 'author',
    }
  }

  async componentDidMount() {
    const { role } = this.state
    const { dispatch } = this.props

    localStorage.setItem('role', role)

    const uri = window.location.href
    const hasCode = uri.includes('?code=')

    if (hasCode) {
      const [url, code] = uri.split('?code=')
      window.history.pushState({}, null, url)
      dispatch(login(code, role))
    }
  }

  handleChangeRole = (role) => {
    localStorage.setItem('role', role)
    this.setState((prev) => ({ ...prev, role }))
  }

  render() {
    const { role } = this.state
    const { isLoggedIn, clientId, redirectURI, loading, errors } = this.props

    if (isLoggedIn) {
      return <Redirect to="/" />
    }

    return (
      <section className="login-form">
        <Card
          style={{ width: 320 }}
          title={<h1 className="login-form__name">X Check App</h1>}
          cover={<img className="login-form__logo" src="./github-logo.png" alt="GitHub" />}
          actions={[
            <>
              {!errors ? (
                <Button
                  type="primary"
                  icon={<GithubOutlined />}
                  loading={loading}
                  href={`https://github.com/login/oauth/authorize?scope=user&client_id=${clientId}&redirect_uri=${redirectURI}`}
                >
                  Sign up with GitHub
                </Button>
              ) : (
                <Alert className="login-form__error" message="Auth wrong." type="error" showIcon />
              )}
            </>,
          ]}
        >
          <Card.Meta
            title="Please login via GitHub"
            description="In order to access the RS School App, you need to login with your GitHub account"
          />
          <div className="select-role">
            <span className="select-role__label">
              Role <span className="req">*</span>
            </span>
            <Select
              defaultValue={role}
              style={{ width: '100%' }}
              onChange={this.handleChangeRole}
              disabled={loading}
            >
              {this.roles.map(({ slug, label }) => (
                <Select.Option value={slug} key={slug}>
                  {label}
                </Select.Option>
              ))}
            </Select>
          </div>
        </Card>
      </section>
    )
  }
}

LoginForm.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  clientId: PropTypes.string.isRequired,
  redirectURI: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired,
  errors: PropTypes.bool.isRequired,
  dispatch: PropTypes.func.isRequired,
}

const mapStateToProps = (state) => {
  return {
    isLoggedIn: state.auth.isLoggedIn,
    clientId: state.auth.clientId,
    redirectURI: state.auth.redirectURI,
    loading: state.auth.loading,
    errors: state.auth.errors,
  }
}

export default connect(mapStateToProps, null)(LoginForm)
