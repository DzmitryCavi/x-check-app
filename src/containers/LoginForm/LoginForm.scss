.login-form {
    text-align: center;
    &__name {
        margin-bottom: 0;
    }
    &__logo {
        display: block;
        width: 100px !important;
        margin: 30px auto 20px;
    }
    &__error {
        text-align: left;
        margin: 0 24px !important;
    }
}

.select-role {
    margin: 20px auto 0;
    text-align: left;
    &__label {
        display: block;
        font-weight: bold;
        font-size: 14px;
    }
}