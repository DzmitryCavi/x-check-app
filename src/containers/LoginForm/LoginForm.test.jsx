import React from 'react'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureStore from 'redux-mock-store'
import LoginForm from './LoginForm'

const mockStore = configureStore([])

const initialState = {
  auth: {
    isLoggedIn: false,
    user: null,
    clientId: 'CLIENT_ID',
    redirectURI: 'REDIRECT_URI',
    clientSecret: 'CLIENT_SECRET',
    proxyURL: 'PROXY_URL',
    loading: false,
    errors: false,
  },
}

const store = mockStore(initialState)

const LoginFormWrapper = (
  <Provider store={store}>
    <LoginForm />
  </Provider>
)

describe('LoginForm component', () => {
  const wrapper = mount(LoginFormWrapper)

  it('Matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Should be default state', () => {
    const LoginFormComponent = wrapper.find('LoginForm')

    expect(LoginFormComponent.state()).toEqual({
      role: 'author',
    })
  })
})
