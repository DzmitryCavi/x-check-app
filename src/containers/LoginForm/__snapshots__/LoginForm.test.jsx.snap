// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm component Matches the snapshot 1`] = `"<section class=\\"login-form\\"><div style=\\"width: 320px;\\" class=\\"ant-card ant-card-bordered\\"><div class=\\"ant-card-head\\"><div class=\\"ant-card-head-wrapper\\"><div class=\\"ant-card-head-title\\"><h1 class=\\"login-form__name\\">X Check App</h1></div></div></div><div class=\\"ant-card-cover\\"><img class=\\"login-form__logo\\" src=\\"./github-logo.png\\" alt=\\"GitHub\\"></div><div class=\\"ant-card-body\\"><div class=\\"ant-card-meta\\"><div class=\\"ant-card-meta-detail\\"><div class=\\"ant-card-meta-title\\">Please login via GitHub</div><div class=\\"ant-card-meta-description\\">In order to access the RS School App, you need to login with your GitHub account</div></div></div><div class=\\"select-role\\"><span class=\\"select-role__label\\">Role <span class=\\"req\\">*</span></span><div class=\\"ant-select ant-select-single ant-select-show-arrow\\" style=\\"width: 100%;\\"><div class=\\"ant-select-selector\\"><span class=\\"ant-select-selection-search\\"><input autocomplete=\\"off\\" class=\\"ant-select-selection-search-input\\" style=\\"opacity: 0;\\" role=\\"combobox\\" aria-haspopup=\\"listbox\\" aria-owns=\\"rc_select_TEST_OR_SSR_list\\" aria-autocomplete=\\"list\\" aria-controls=\\"rc_select_TEST_OR_SSR_list\\" aria-activedescendant=\\"rc_select_TEST_OR_SSR_list_0\\" readonly=\\"\\" unselectable=\\"on\\" value=\\"\\" id=\\"rc_select_TEST_OR_SSR\\"></span><span class=\\"ant-select-selection-item\\" title=\\"Author\\">Author</span></div><span class=\\"ant-select-arrow\\" style=\\"user-select: none;\\" unselectable=\\"on\\" aria-hidden=\\"true\\"><span role=\\"img\\" aria-label=\\"down\\" class=\\"anticon anticon-down ant-select-suffix\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"down\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z\\"></path></svg></span></span></div></div></div><ul class=\\"ant-card-actions\\"><li style=\\"width: 100%;\\"><span><a href=\\"https://github.com/login/oauth/authorize?scope=user&amp;client_id=CLIENT_ID&amp;redirect_uri=REDIRECT_URI\\" class=\\"ant-btn ant-btn-primary\\"><span role=\\"img\\" aria-label=\\"github\\" class=\\"anticon anticon-github\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"github\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9a127.5 127.5 0 0138.1 91v112.5c.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z\\"></path></svg></span><span>Sign up with GitHub</span></a></span></li></ul></div></section>"`;
